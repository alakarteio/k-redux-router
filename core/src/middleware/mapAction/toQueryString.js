// https://stackoverflow.com/questions/1714786/query-string-encoding-of-a-javascript-object
export default obj => Object.keys(obj).map(k => `${encodeURIComponent(k)}=${encodeURIComponent(obj[k])}`).join('&')
