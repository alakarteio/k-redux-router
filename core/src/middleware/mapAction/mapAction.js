/* eslint-env browser */
import routeFound from '../routeFound'
import toQueryString from './toQueryString'

export default (options, reducer) => {
  const {
    history,
  } = options

  return store => (action) => {
    const { type, payload } = action

    switch (type) {
      case '@@router/GO_FORWARD': {
        history.go(payload)
        return undefined
      }
      case '@@router/GO_BACK': {
        history.go(payload * -1)
        return undefined
      }
      case '@@router/REPLACE':
      case '@@router/PUSH': {
        const { code, params = {} } = payload

        // find route
        const route = reducer.getState(store.getState()).routes.map[code]

        if (route) {
          const { href } = route

          // remove unknown path params
          const pathParamsToPush = {}
          if (params && params.path && href.parsed) {
            href.parsed
              .filter(part => typeof part === 'object')
              .map(part => part.name)
              .filter(name => params.path[name])
              .forEach((name) => { pathParamsToPush[name] = params.path[name] })
          }

          // construct URI
          let queryPart = ''
          let toPush = href.base
          if (href.compiled) toPush = href.compiled(pathParamsToPush)
          if (params.query) queryPart = `?${toQueryString(params.query)}`

          // push URI to history API
          toPush = `${toPush}${queryPart}`
          if (type === '@@router/PUSH') history.pushState(undefined, undefined, toPush)
          else history.replaceState(undefined, undefined, toPush)


          // update state
          const paramsToPush = { ...params }
          if (params && params.path) paramsToPush.path = pathParamsToPush
          return routeFound({
            route,
            params: paramsToPush,
          })
        }

        // TODO: if no route, find the closest `notFound` to push it in `result`
        return undefined
      }
      // default is undefined so we avoid infinite loop
      default: return undefined
    }
  }
}
