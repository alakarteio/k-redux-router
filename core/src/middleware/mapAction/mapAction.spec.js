/* eslint-env jest */
import mapAction from './mapAction'

describe('mapAction', () => {
  const options = {
    history: {
      go: jest.fn(),
      pushState: jest.fn(),
      replaceState: jest.fn(),
    },
  }

  const reducer = {
    getState: state => state.router,
  }

  const state = {
    router: {
      routes: {
        map: {
          'found-code': {
            href: {
              base: '/route/:id/with/parameters',
              compiled: jest.fn(() => 'compiled/route/'),
            },
          },
          'found-code-2': {
            href: {
              base: '/route/without/parameters',
            },
          },
        },
      },
    },
  }

  const store = {
    getState: () => state,
  }

  beforeEach(() => {
    options.history.go.mockClear()
    options.history.pushState.mockClear()
    options.history.replaceState.mockClear()
    state.router.routes.map['found-code'].href.compiled.mockClear()
  })

  const match = action => () => expect({
    returns: mapAction(options, reducer)(store)(action),
    go: options.history.go.mock.calls,
    pushState: options.history.pushState.mock.calls,
    replaceState: options.history.replaceState.mock.calls,
    compiled: state.router.routes.map['found-code'].href.compiled.mock.calls,
  }).toMatchSnapshot()

  // !!! TESTS !!!
  it('should return undefined when action is unknown', match({ type: 'UNKNOWN' }))
  it('should go forward and return undefined action', match({ type: '@@router/GO_FORWARD', payload: 3 }))
  it('should go backward and return undefined action', match({ type: '@@router/GO_BACK', payload: 5 }))
  it('should return a ROUTE_FOUND action with compiled params -push-', match({
    type: '@@router/PUSH',
    payload: {
      code: 'found-code',
      params: {
        path: { one: 'path-param' },
        query: { an: 'other-query-param' },
      },
    },
  }))
  it('should return a ROUTE_FOUND action with compiled params -replace-', match({
    type: '@@router/REPLACE',
    payload: {
      code: 'found-code-2',
    },
  }))
})
