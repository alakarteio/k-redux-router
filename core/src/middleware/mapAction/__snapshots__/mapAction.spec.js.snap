// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapAction should go backward and return undefined action 1`] = `
Object {
  "compiled": Array [],
  "go": Array [
    Array [
      -5,
    ],
  ],
  "pushState": Array [],
  "replaceState": Array [],
  "returns": undefined,
}
`;

exports[`mapAction should go forward and return undefined action 1`] = `
Object {
  "compiled": Array [],
  "go": Array [
    Array [
      3,
    ],
  ],
  "pushState": Array [],
  "replaceState": Array [],
  "returns": undefined,
}
`;

exports[`mapAction should return a ROUTE_FOUND action with compiled params -push- 1`] = `
Object {
  "compiled": Array [
    Array [
      Object {
        "one": "path-param",
      },
    ],
  ],
  "go": Array [],
  "pushState": Array [
    Array [
      undefined,
      undefined,
      "compiled/route/?an=other-query-param",
    ],
  ],
  "replaceState": Array [],
  "returns": Object {
    "payload": Object {
      "found": true,
      "params": Object {
        "path": Object {
          "one": "path-param",
        },
        "query": Object {
          "an": "other-query-param",
        },
      },
      "route": Object {
        "href": Object {
          "base": "/route/:id/with/parameters",
          "compiled": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "one": "path-param",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": "compiled/route/",
              },
            ],
          },
          "parsed": Array [
            "/something",
            Object {
              "name": "one",
            },
          ],
        },
      },
    },
    "type": "@@router/ROUTE_FOUND",
  },
}
`;

exports[`mapAction should return a ROUTE_FOUND action with compiled params -replace- 1`] = `
Object {
  "compiled": Array [],
  "go": Array [],
  "pushState": Array [],
  "replaceState": Array [
    Array [
      undefined,
      undefined,
      "/route/without/parameters",
    ],
  ],
  "returns": Object {
    "payload": Object {
      "found": true,
      "params": Object {},
      "route": Object {
        "href": Object {
          "base": "/route/without/parameters",
        },
      },
    },
    "type": "@@router/ROUTE_FOUND",
  },
}
`;

exports[`mapAction should return undefined when action is unknown 1`] = `
Object {
  "compiled": Array [],
  "go": Array [],
  "pushState": Array [],
  "replaceState": Array [],
  "returns": undefined,
}
`;
